### PR TITLE
[2.7][Translation] added message cache.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | #13948, #2570, #13676 |
| Tests pass? | yes |
| License | MIT |
- [x] tests
- [x] add MessageDoctrineCache

``` yml
# app/config/config.yml
framework:
    translator:
        cache: translation.cache.doctrine # translation.cache.default
```

### CODE REVIEW

File: src/Symfony/Component/Translation/Translator.php [hunk 1]
```diff
@@ -76,11 +70,15 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
      *
      * @api
      */
-    public function __construct($locale, MessageSelector $selector = null, $cacheDir = null, $debug = false)
+    public function __construct($locale, MessageSelector $selector = null, $cache = null, $debug = false)
```
Comment (xabbuh) [inline src/Symfony/Component/Translation/Translator.php]: The docblock needs to be changed to take into account that `$cache` can be an instance of `MessageCacheInterface`.

**Comment by aitboudad** [comment]
I think this could be a nice addition for 2.7, the default cache is useful in 80% of apps but when the size of catalogue became more than 10 MB it's consumes too much memory.
So the solution is to use DoctrineMessageCache, I tested with a catalogue(~100 mb) and the result look really interesting:

```yaml
#config.yml
framework:
    translator:      { fallbacks: [en], cache: translation.cache.doctrine }

doctrine_cache:
    providers:
        translation_cache:
            memcached:
                servers:
                    localhost: 11211
```
```yaml
#services.yml
translation.cache.doctrine:
    class: Symfony\Component\Translation\DoctrineMessageCache
    arguments: [%kernel.debug%, @translation_cache]
```

https://blackfire.io/profiles/compare/82c496fe-200a-451d-a2c8-e2fd458f0193/graph
selection_012
 
/cc @fabpot

File: src/Symfony/Component/Translation/Translator.php (hunk 1 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Translator.php]: fixed

**Comment by aitboudad** [comment]
@Tobion can you review this PR ?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 2]
```diff
@@ -152,5 +153,13 @@
         <service id="translation.extractor" class="%translation.extractor.class%"/>
 
         <service id="translation.writer" class="%translation.writer.class%"/>
+
+        <!-- Cache -->
+        <service id="translation.cache.default" class="Symfony\Component\Translation\MessageCache" public="false">
```
Comment (GromNaN) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: You can mâle this service private. Then iT could be inlined or removed by thé compiler.

Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: isn't it ?

Comment (GromNaN) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: Yes it is already. Sorry.

**Comment by aitboudad** [comment]
ping @symfony/deciders

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -652,6 +652,10 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
 
         $container->setParameter('translator.logging', $config['logging']);
 
+        if (isset($config['cache']) && $config['cache']) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Isn't `isset($config['cache'])` alone sufficient enough?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 4]
```diff
@@ -152,5 +153,13 @@
...
+            <argument key="cache_dir">%kernel.cache_dir%/translations</argument> <!-- cache_dir -->
+            <argument>%kernel.debug%</argument> <!-- debug -->
+        </service>
+
+        <service id="translation.cache" alias="translation.cache.default" />
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: Do we need this? We already set the alias in the extension class. And in case the alias is not set (e.g. because the option is not used), we can set `on-invalid="null"` on the `translation.cache` argument in line 48.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php [hunk 5]
```diff
@@ -146,6 +146,7 @@ protected static function getBundleDefaultConfig()
                 'enabled' => false,
                 'fallbacks' => array('en'),
                 'logging' => true,
+                'cache' => 'translation.cache.default',
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php]: We should probably also add a test where this is explicitly set to the empty string. And what about testing it with an invalid value?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml (hunk 4 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: nice one, I used ``` on-invalid="null" ``` thanks

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php (hunk 5 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php]: We will just duplicate tests if we add these case, as it already covered by the container.

**Comment by aitboudad** [comment]
any feedback, I really want it to be available in 2.7 :)

**Comment by xabbuh** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 6]
```diff
@@ -152,5 +153,11 @@
         <service id="translation.extractor" class="%translation.extractor.class%"/>
 
         <service id="translation.writer" class="%translation.writer.class%"/>
+
+        <!-- Cache -->
+        <service id="translation.cache.default" class="Symfony\Component\Translation\MessageCache" public="false">
+            <argument key="cache_dir">%kernel.cache_dir%/translations</argument> <!-- cache_dir -->
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: The `key` can be removed as it's not used.

Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: fixed

File: src/Symfony/Component/Translation/Translator.php [hunk 7]
```diff
@@ -55,32 +54,29 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
...
-    /**
-     * Constructor.
-     *
-     * @param string               $locale   The locale
-     * @param MessageSelector|null $selector The message selector for pluralization
-     * @param string|null          $cacheDir The directory to use for the cache
-     * @param bool                 $debug    Use cache in debug mode ?
+     * @param string                            $locale   The locale
+     * @param MessageSelector|null              $selector The message selector for pluralization
+     * @param string|null|MessageCacheInterface $cache    The directory to use for the cache
```
Comment (stof) [inline src/Symfony/Component/Translation/Translator.php]: the description is not accurate anymore

File: src/Symfony/Component/Translation/DoctrineMessageCache.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+namespace Symfony\Component\Translation;
+
+use Doctrine\Common\Cache\Cache;
+
+/**
+ * @author Abdellatif Ait Boudad <a.aitboudad@gmail.com>
+ */
+class DoctrineMessageCache implements MessageCacheInterface
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCache.php]: This class should go in the Doctrine bridge

File: src/Symfony/Component/Translation/DoctrineMessageCache.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        $this->cacheProvider = $cacheProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isFresh($locale, $options = array())
+    {
+        if (!isset($options['resources_hash'])) {
+            $options['resources_hash'] = '';
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCache.php]: instead of altering the array of options to add the key just to be able to read it later, I would rather use a local variable for the expected resource hash.
this will avoid copying the array because of writing in it

File: src/Symfony/Component/Translation/DoctrineMessageCatalogue.php [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+namespace Symfony\Component\Translation;
+
+use Doctrine\Common\Cache\Cache;
+use Doctrine\Common\Cache\MultiGetCache;
+
+/**
+ * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
+ */
+class DoctrineMessageCatalogue extends MessageCatalogue
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCatalogue.php]: should go in the bridge too

File: src/Symfony/Component/Translation/DoctrineMessageCatalogue.php [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+     */
+    public function __construct($locale, Cache $cache, $prefix = self::PREFIX)
+    {
+        parent::__construct($locale);
+        if (0 === strlen($prefix)) {
+            throw new \InvalidArgumentException('$prefix cannot be empty.');
+        }
+
+        $this->cache = $cache;
+        $this->prefix = $prefix.$locale;
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCatalogue.php]: any place using the prefix is adding a ``_`` after it. I suggest adding it here already

File: src/Symfony/Component/Translation/DoctrineMessageCatalogue.php [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        $domains = $this->domains;
+        if (null !== $domain) {
+            $domains = array($domain);
+        }
+
+        $messages = array();
+        foreach ($domains as $domainMeta) {
+            $domainIdentity = $this->prefix.'_'.self::CATALOGUE_DOMAIN_METADATA.'_'.$domainMeta;
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCatalogue.php]: no need to add the ``_`` after ``self::CATALOGUE_DOMAIN_METADATA``. It is already in the constant

File: src/Symfony/Component/Translation/DoctrineMessageCache.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * @var bool
+     */
+    private $debug;
+
+    /**
+     * @var Cache
+     */
+    private $cacheProvider;
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCache.php]: I would rename the property to ``cache``

File: src/Symfony/Component/Translation/DoctrineMessageCache.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function dump(MessageCatalogueInterface $messages, $options = array())
+    {
+        while ($messages) {
+            $catalogue = new DoctrineMessageCatalogue($messages->getLocale(), $this->cacheProvider);
+            $catalogue->addCatalogue($messages);
+
+            $this->cacheProvider->save($this->getResourceHashKey($messages->getLocale()), $options['resources_hash']);
```
Comment (stof) [inline src/Symfony/Component/Translation/DoctrineMessageCache.php]: what if ``$options['resources_hash']`` is not defined ?

File: src/Symfony/Component/Translation/MessageCache.php [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        $this->debug = $debug;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isFresh($locale, $options = array())
+    {
+        if (!isset($options['resources_hash'])) {
+            $options['resources_hash'] = '';
```
Comment (stof) [inline src/Symfony/Component/Translation/MessageCache.php]: same here abotu avoiding to write in the array

File: src/Symfony/Component/Translation/MessageCache.php [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load($locale)
+    {
+        $resource = $this->getResourcePath($locale);
+
+        list($catalogue, $resourcesHash) = include $resource;
+        $catalogue->addResource(new FileResource($resource));
```
Comment (stof) [inline src/Symfony/Component/Translation/MessageCache.php]: adding the cache file as a resource looks weird to me. Are you sure about that ?

File: src/Symfony/Component/Translation/Translator.php [hunk 17]
```diff
@@ -320,7 +316,12 @@ public function getMessages($locale = null)
      */
     protected function loadCatalogue($locale)
     {
-        if (null === $this->cacheDir) {
+        $this->assertValidLocale($locale);
```
Comment (stof) [inline src/Symfony/Component/Translation/Translator.php]: do we need to assert it here ? We could do the validation only when we don't have the catalogue yet (as currently), thanks to the fact that we will never create ``$this->catalogues[$locale]`` for an invalid locale

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Bridge/Doctrine/Translation/DoctrineMessageCatalogue.php [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        parent::__construct($locale);
+        if (0 === strlen($prefix)) {
+            throw new \InvalidArgumentException('$prefix cannot be empty.');
+        }
+
+        $this->cache = $cache;
+        $this->prefix = $prefix.'_'.$locale;
+
+        if ($cache->contains($domainsId = $this->prefix.'_'.self::CATALOGUE_DOMAINS)) {
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Translation/DoctrineMessageCatalogue.php]: This ``_`` can be added when creating ``$this->prefix`` instead of doing it for each usage (as said previously)

**Comment by aitboudad** [comment]
@stof comments fixed, thanks :)
For doc: https://github.com/symfony/symfony-docs/issues/5135

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 19]
```diff
@@ -659,6 +659,10 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
 
         $container->setParameter('translator.logging', $config['logging']);
 
+        if (isset($config['cache'])) {
+            $container->setAlias('translation.cache', $config['cache']);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: should be ``translator.cache`` to be consistent with our other ids

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: hmm, no. We are already using both actually

File: src/Symfony/Component/Translation/MessageCacheInterface.php [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /**
+     * Returns true if the cache is still fresh.
+     *
+     * @param string $locale
+     * @param array  $options
+     *
+     * @return bool
+     */
+    public function isFresh($locale, $options = array());
```
Comment (stof) [inline src/Symfony/Component/Translation/MessageCacheInterface.php]: the ``options`` argument should be typehinted (same in ``dump()``)

Comment (aitboudad) [inline src/Symfony/Component/Translation/MessageCacheInterface.php]: fixed

**Comment by aitboudad** [comment]
can I merge ?

**Comment by mpdude** [comment]
This PR basically adds a new concept similar to `ConfigCache`, but explicitly for messages. It moves us away from a consistent application of `ConfigCache`.

Once this is merged (and becomes subject to the BC promise), I don't know if we can follow the ideas of #14178 (ex #7781) or service-based resource validation (#7230) anymore.

**Comment by fabpot** [comment]
I tend to agree with @mpdude

**Comment by aitboudad** [comment]
anyway the main thing here is to provide a doctrine cache for translation, what I didn't like for ConfigCache is it mainly related to resource file.

**Comment by aitboudad** [comment]
any decision about this feature the current implementation look good to me :),
the #14178 look nice but didn't solve the issue at all,
another thing to consider here is that users can use the cache without the config component deps.

**Comment by mpdude** [comment]
#14178 tries to solve a completely different problem :-)

**Comment by mpdude** [comment]
If I got the code right (not sure about that), your goal is to keep messages in a key-value-store type cache. 

That way, the `Translator` does not have to load a possibly huge `MessageCatalogue` (or even several of them) from the `ConfigCache` only to return a single message. It can possibly retrieve the message directly from the key-value-cache.

Also, the message catalogue will mostly be kept in shared memory and reduce the memory footprint for a single PHP process. Currently the entire Catalogue will be read from the `ConfigCache` into per-request memory.

So, if this understanding is correct, can't we find a way to make the Translator (key-value-)cache aware and use a lookup strategy based on (locale, domain, message-id) without having to touch the `ConfigCache` related parts at all?

Here's the steps such a Translator would need to do:

1. Figure out if the key-value-cache is fresh.

For this, we could stick with the current `ConfigCache` validation. If we detect that we need to rebuild the `ConfigCache`, we subsequently also need to copy/map the Catalogue (after we read it) into the key-value-store/cache.

2. Look into the cache

Before `require`ing the Catalogue from the `ConfigCache`, check whether the requested (locale, domain, message-id) is available in the cache. If so, return it from there.

3. Cache miss

If a requested translation cannot be found, it does either not exist or it may be missing because the key-value-cache has been purged. 

We could detect the latter by having an extra key that is written once the cache was loaded. If that key is missing, re-populate the key-value-cache with the `MessageCatalogue` loaded from the `ConfigCache` (!).

Does that make sense?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 21]
```diff
@@ -659,6 +659,10 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
 
         $container->setParameter('translator.logging', $config['logging']);
 
+        if (isset($config['cache'])) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: You should prefix cache entries with something specific with the current installation to prevent conflicts if several Symfony apps run on the same server.

The example for the validator: https://github.com/aitboudad/symfony/blob/translation_catalogue_cache/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L774

File: src/Symfony/Component/Translation/MessageCache.php [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+<?php
+use Symfony\Component\Translation\MessageCatalogue;
+\$resourcesHash = '%s';
+\$catalogue = new MessageCatalogue('%s', %s);
+%s
+return array(\$catalogue, \$resourcesHash);
+EOF
+            ,
+            $resourcesHash,
+            $messages->getLocale(),
```
Comment (dunglas) [inline src/Symfony/Component/Translation/MessageCache.php]: Is it ok in term of security? Are we sure those values never come from an end user (I think about complex "community-driven" translation workflows).

**Comment by aitboudad** [comment]
@mpdude you understand this topic now better than me :) , below my answers:

>If I got the code right (not sure about that), your goal is to keep messages in a key-value-store type cache.

The main think here for me is to move the cache outside of Translator and adding an initial implementation of key-value-store(doctrine/cache).

> can't we find a way to make the Translator (key-value-)cache aware 
Right I think missed this part even if we use key-value-store we need ```ConfigCache`` to make cache aware better.
I work now on this part so I'll fix it soon.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 21 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: fixed.

**Comment by aitboudad** [comment]
@mpdude Hi,
If we decide to use the Translator (key-value-) we don't need anymore to save cache in file !! and I think the current limitations of  ```ConfigCache``` is it doesn't support key/value cache.

So what I made now with 69225b83addd464a8a8351176ed449bcf5275254 is refresh cache when resources file change.

And to understand how I make the Translator (key-value-)cache aware:
 
>1 - Figure out if the catalogue is still fresh/valid.

[isFresh](https://github.com/aitboudad/symfony/blob/translation_catalogue_cache/src/Symfony/Bridge/Doctrine/Translation/DoctrineMessageCache.php#L51)

>2 - Look into the cache

[check whether the requested (locale, domain, message-id) is available in the (key-value-)cache](https://github.com/aitboudad/symfony/blob/translation_catalogue_cache/src/Symfony/Bridge/Doctrine/Translation/DoctrineMessageCatalogue.php#L133-L138)

>3 - Cache miss

If that key is missing, re-populate the key-value-cache with the MessageCatalogue loaded from the Loaders and not ConfigCache !

File: src/Symfony/Component/Translation/MessageCache.php (hunk 22 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/MessageCache.php]: I agree with you :), an end user can inject PHP code into messages !! but I'm not sure if [var_export](https://github.com/aitboudad/symfony/blob/translation_catalogue_cache/src/Symfony/Component/Translation/MessageCache.php#L91) solve the issue or not ? //cc @fabpot

**Comment by mpdude** [comment]
If you're trying to go without `ConfigCache`, you will end up re-implementing the freshness check it provides (for example [here](https://github.com/aitboudad/symfony/commit/69225b83addd464a8a8351176ed449bcf5275254#diff-8f5d9bb6230224897834742b0d19c8a8R113)). You also seem need the `DoctrineMessageCache` as a replacement for `ConfigCache`.

My suggestion was to keep the `ConfigCache` part as-is. It's just a file sitting on the disk and does no harm. That way, you can still use the resource validation logic it provides. You will see from within `Translator` when the catalogue has to be rebuilt, that's when the cache-is-fresh check fails.

In other words, the `DoctrineCache` would just be another caching layer on top of `ConfigCache`: In `Translator`, before you load the entire Catalogue from disk, see if you can find that particular message in the DoctrineCache first.

When you have the Catalogue at hand anyways (because you had to rebuild it *or* load it because the message was not in DoctrineCache), also save it to the DoctrineCache.

**Comment by aitboudad** [comment]
@mpdude well I agree to some extent, but I didn't like the idea of writing cache twice in ConfigCache then DoctrineCache.

@fabpot what do you think?

**Comment by jakzal** [comment]
I'd hold off and do not merge this into 2.7. Note that I didn't review this PR in details just yet, and I only scanned through comments. Forgive me if I'm repeating someone else here.

We need to think of how to organise doctrine integration more carefully and make it consistent across the framework. With the current implementation I don't think that doctrine cache classes belong to the Doctrine bridge, as they're only use is the translator.

Serializer and Validator have their own cache implementations, slightly different from each other - first relies on doctrine/cache directly, the later adapts its interface. This PR introduces yet another approach. All I'm saying we should stop and consider if we can make it a bit more consistent.

**Comment by aitboudad** [comment]
The safe way is to split this PR into 2, let's discuss about adding [message cache interface](#14526) and once accepted I'll open another one for doctrine cache.
